### PR TITLE
add knn.ipynb and update other notebooks

### PR DESCRIPTION
**Pull request by leewyang**

### CODE REVIEW

**Comment by leewyang** [comment]
build

File: notebooks/knn.ipynb [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+    "# Compute the locality sensitive hashes for the input rows, then perform approximate similarity join.\n",
+    "# We could avoid computing hashes by passing in the already-transformed dataset, e.g.\n",
+    "# `model.approxSimilarityJoin(transformedA, transformedB, 1.5)`\n",
+    "print(\"Approximately joining items and queries on Euclidean distance smaller than 3.0:\")\n",
+    "model.approxSimilarityJoin(item_vector_df, query_vector_df, 3.0, distCol=\"EuclideanDistance\")\\\n",
+    "    .select(col(\"datasetA.features\").alias(\"item\"),\n",
+    "            col(\"datasetB.features\").alias(\"query\"),\n",
+    "            col(\"EuclideanDistance\")).orderBy(\"query\", \"item\").show()\n",
+    "\n",
+    "# results = model.approxSimilarityJoin(item_vector_df, query_vector_df, 1.5, distCol=\"EuclideanDistance\")"
```
Comment (eordentlich) [inline notebooks/knn.ipynb:345]: Why the commented line here?

File: notebooks/knn.ipynb [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+   "execution_count": null,
+   "id": "e70d24b7-db70-4041-9794-7cd5451bad76",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "# Compute the locality sensitive hashes for the input rows, then perform approximate nearest neighbor search.\n",
+    "# We could avoid computing hashes by passing in the already-transformed dataset, e.g.\n",
+    "# `model.approxNearestNeighbors(transformedA, key, 2)`\n",
```
Comment (eordentlich) [inline notebooks/knn.ipynb:359]: was there a transformedA earlier?

File: notebooks/knn.ipynb [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "8cd56670-7633-4fe6-ab75-0fd680c63baa",
+   "metadata": {},
+   "source": [
+    "# PySpark\n",
+    "\n",
+    "Note: PySpark does not have an exact kNN implementation, but it does have an LSH-based approximate nearest neighbors implementation, shown here."
```
Comment (eordentlich) [inline notebooks/knn.ipynb:254]: Maybe add a clarifying comment about why this is here.  Like to illustrate similarity of apis.   Ideally the results could be compared but that would require a more elaborate tuning of parameters here.

File: notebooks/knn.ipynb [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "import numpy as np\n",
+    "\n",
+    "from spark_rapids_ml.knn import NearestNeighbors\n",
+    "from pyspark.sql import DataFrame, Row\n",
+    "from pyspark.sql.functions import col, lit\n",
+    "from pyspark.sql.types import IntegerType, StructType, StructField"
```
Comment (lijinf2) [inline notebooks/knn.ipynb:25]: It seems "lit", "IntegerType", "StructType", and "StructField" do not appear in other places of the notebook. 
May need to delete them if not used. 

It seems "DataFrame" and "Row" are also not used.

File: notebooks/spark-compat.ipynb [hunk 5]
```diff
@@ -84,7 +84,21 @@
...
      "text": [
-      "                                                                                \r"
+      "[Stage 0:>                                                          (0 + 0) / 1]\r"
+     ]
+    },
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "23/04/04 16:27:16 WARN TaskSchedulerImpl: Initial job has not accepted any resources; check your cluster UI to ensure that workers are registered and have sufficient resources\n"
```
Comment (lijinf2) [inline notebooks/spark-compat.ipynb:94]: Is this warning expected here?

File: notebooks/knn.ipynb (hunk 2 above)
Comment (lijinf2) [reply notebooks/knn.ipynb:359]: transformedA seems a term used to comment CPU example codes: https://spark.apache.org/docs/2.2.3/ml-features.html#lsh-operations.  It is not a variable.

**Comment by lijinf2** [comment]
Minor things. Mostly reads good.

File: notebooks/knn.ipynb [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "0854c26e-b4f3-480b-b8c2-ab6ae4f73d86",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "item = [\n",
```
[the remark this answers is not in the document]
Comment (leewyang) [reply notebooks/knn.ipynb]: Updated w/ randomly generated dataset.

File: notebooks/knn.ipynb [hunk 7]
```diff
@@ -0,0 +1,399 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "21d0f88c-8ba0-4f25-a42f-45c864972f54",
+   "metadata": {},
+   "source": [
+    "#### exactNearestNeighborsJoin\n",
+    "\n",
+    "This API returns a join of the query vectors and their `k` closest item vectors."
```
[the remark this answers is not in the document]
Comment (leewyang) [reply notebooks/knn.ipynb:252]: Done.

File: notebooks/knn.ipynb (hunk 1 above)
Comment (leewyang) [reply notebooks/knn.ipynb]: Whoops, removed.

File: notebooks/knn.ipynb (hunk 2 above)
Comment (leewyang) [reply notebooks/knn.ipynb]: Fixed to reflect variables in notebook.

File: notebooks/knn.ipynb (hunk 3 above)
Comment (leewyang) [reply notebooks/knn.ipynb]: Done.

File: notebooks/knn.ipynb (hunk 4 above)
Comment (leewyang) [reply notebooks/knn.ipynb]: Good catch, removed.

File: notebooks/spark-compat.ipynb (hunk 5 above)
Comment (leewyang) [reply notebooks/spark-compat.ipynb]: Removed.

**Review by lijinf2 — APPROVED** [review]
Looks nice!

**Comment by lijinf2** [comment]
build

File: notebooks/knn.ipynb [hunk 8]
```diff
@@ -0,0 +1,447 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "38c671ba-b6de-4414-992f-957e28f3a8be",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "dim = 2000\n",
```
Comment (eordentlich) [inline notebooks/knn.ipynb:44]: Does this work on Databricks?

Comment (leewyang) [reply notebooks/knn.ipynb:44]: Just tried.  Yes, although it looks like there might be a bug w/ `exactNearestNeighborsJoin` returning no results (even at smaller data sizes).

**Pull request merged**